File: LabGen/labgen.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.Linq;

class MazeGenerator
{
    static void Main(string[] args)
    {
        int width = 21, height = 21; // Dimensions must be odd
        int cellSize = 20; // Size of each cell in the image

        int[,] maze = GenerateMaze(width, height);
        List<(int, int)> path = FindShortestPath(maze);

        DrawMazeWithPath(maze, path, cellSize, @"C:\Users\Vlad\Pictures\maze_with_path.png");
        Console.WriteLine("Maze with path saved to C:\\Users\\Vlad\\Pictures\\maze_with_path.png");
    }

    // Generates a random maze using depth-first search
    static int[,] GenerateMaze(int width, int height)
    {
        int[,] maze = new int[height, width];
        for (int y = 0; y < height; y++)
            for (int x = 0; x < width; x++)
                maze[y, x] = 1; // Fill with walls

        Stack<(int, int)> stack = new();
        Random rand = new();
        maze[1, 1] = 0;
        stack.Push((1, 1));

        int[] dx = { 0, 0, 2, -2 };
        int[] dy = { 2, -2, 0, 0 };

        while (stack.Count > 0)
        {
            var (x, y) = stack.Peek();
            List<int> directions = Enumerable.Range(0, 4).OrderBy(_ => rand.Next()).ToList();
            bool moved = false;

            foreach (int dir in directions)
            {
                int nx = x + dx[dir], ny = y + dy[dir];
                int wallX = x + dx[dir] / 2, wallY = y + dy[dir] / 2;

                if (nx > 0 && ny > 0 && nx < height - 1 && ny < width - 1 && maze[nx, ny] == 1)
                {
                    maze[wallX, wallY] = 0; // Remove wall
                    maze[nx, ny] = 0; // Mark as path
                    stack.Push((nx, ny));
                    moved = true;
                    break;
                }
            }

            if (!moved)
                stack.Pop(); // Backtrack
        }

        return maze;
    }

    // Finds the shortest path from (1, 1) to (height-2, width-2)
    static List<(int, int)> FindShortestPath(int[,] maze)
    {
        int height = maze.GetLength(0), width = maze.GetLength(1);
        Queue<(int x, int y, List<(int, int)> path)> queue = new();
        HashSet<(int, int)> visited = new();

        queue.Enqueue((1, 1, new List<(int, int)> { (1, 1) }));

        while (queue.Count > 0)
        {
            var (x, y, path) = queue.Dequeue();

            if (x == height - 2 && y == width - 2)
                return path;

            if (visited.Contains((x, y)))
                continue;

            visited.Add((x, y));

            foreach (var (dx, dy) in new[] { (0, 1), (1, 0), (0, -1), (-1, 0) })
            {
                int nx = x + dx, ny = y + dy;
                if (nx > 0 && ny > 0 && nx < height && ny < width && maze[nx, ny] == 0)
                {
                    var newPath = new List<(int, int)>(path) { (nx, ny) };
                    queue.Enqueue((nx, ny, newPath));
                }
            }
        }

        return new List<(int, int)>();
    }

    // Draws the maze and the shortest path, and saves it as an image
    static void DrawMazeWithPath(int[,] maze, List<(int, int)> path, int cellSize, string savePath)
    {
        int width = maze.GetLength(1), height = maze.GetLength(0);
        using Bitmap bmp = new(width * cellSize, height * cellSize);
        using Graphics g = Graphics.FromImage(bmp);

        g.Clear(Color.White);

        // Draw maze
        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                if (maze[y, x] == 1)
                {
                    g.FillRectangle(Brushes.Black, x * cellSize, y * cellSize, cellSize, cellSize);
                }
            }
        }

        // Draw path
        using Pen pen = new(Color.Red, 3);
        for (int i = 0; i < path.Count - 1; i++)
        {
            var (x1, y1) = path[i];
            var (x2, y2) = path[i + 1];
            g.DrawLine(pen,
                x1 * cellSize + cellSize / 2, y1 * cellSize + cellSize / 2,
                x2 * cellSize + cellSize / 2, y2 * cellSize + cellSize / 2);
        }

        // Save the image
        bmp.Save(savePath, ImageFormat.Png);
    }
}

File: LabGen/labgen.py
import random
from PIL import Image, ImageDraw

def generate_maze(width, height):
    # Initialize the maze with walls (1)
    maze = [[1 for _ in range(width)] for _ in range(height)]
    stack = [(1, 1)]  # Start position
    maze[1][1] = 0  # Mark start as path

    # Directions for carving
    directions = [(0, 2), (0, -2), (2, 0), (-2, 0)]

    while stack:
        x, y = stack[-1]
        random.shuffle(directions)  # Shuffle directions for randomness
        moved = False

        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            wall_x, wall_y = x + dx // 2, y + dy // 2
            if 1 <= nx < height - 1 and 1 <= ny < width - 1 and maze[nx][ny] == 1:
                maze[wall_x][wall_y] = 0  # Remove wall
                maze[nx][ny] = 0  # Mark cell as path
                stack.append((nx, ny))
                moved = True
                break

        if not moved:
            stack.pop()  # Backtrack if no moves are possible

    return maze

def find_shortest_path(maze):
    height, width = len(maze), len(maze[0])
    start = (1, 1)
    end = (height - 2, width - 2)
    queue = [(start, [start])]  # (current position, path taken)
    visited = set()

    while queue:
        (x, y), path = queue.pop(0)
        if (x, y) == end:
            return path  # Return the shortest path
        if (x, y) in visited:
            continue
        visited.add((x, y))

        # Add neighbors (up, down, left, right)
        for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            nx, ny = x + dx, y + dy
            if 0 <= nx < height and 0 <= ny < width and maze[nx][ny] == 0:
                queue.append(((nx, ny), path + [(nx, ny)]))

    return []  # No path found

def draw_maze_with_path(maze, path, cell_size=20):
    height, width = len(maze), len(maze[0])
    img_width = width * cell_size
    img_height = height * cell_size

    # Create a new image
    img = Image.new("RGB", (img_width, img_height), "white")
    draw = ImageDraw.Draw(img)

    # Draw the maze
    for y, row in enumerate(maze):
        for x, cell in enumerate(row):
            color = "black" if cell == 1 else "white"
            top_left = (x * cell_size, y * cell_size)
            bottom_right = ((x + 1) * cell_size, (y + 1) * cell_size)
            draw.rectangle([top_left, bottom_right], fill=color)

    # Draw the path with red arrows
    for (x1, y1), (x2, y2) in zip(path, path[1:]):
        start = (y1 * cell_size + cell_size // 2, x1 * cell_size + cell_size // 2)
        end = (y2 * cell_size + cell_size // 2, x2 * cell_size + cell_size // 2)
        draw.line([start, end], fill="red", width=3)

    return img

# Parameters
width, height = 33, 17  # Must be odd numbers

# Generate the maze and find the path
maze = generate_maze(width, height)
shortest_path = find_shortest_path(maze)

# Draw the maze with the optimal path
maze_image = draw_maze_with_path(maze, shortest_path, cell_size=20)

# Save the image to C:\Users\Vlad\Pictures
save_path = r"C:\Users\Vlad\Pictures\maze_with_path.png"
maze_image.save(save_path)
print(f"Maze with path saved to {save_path}")

# Optionally display the image
maze_image.show()
